<a href="https://colab.research.google.com/github/ArshockAbedan/ML2021/blob/main/NLP_BRET.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**BERT tutorial**

---




In [1]:
import tensorflow_hub as hub
import tensorflow_text as text

In [2]:
preprocess_url = "https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3"
encoder_url = "https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4"

In [3]:
bert_preprocess_model = hub.KerasLayer(preprocess_url)

In [4]:
text_test = ['My Name is Aref Abedjooy','I love NLP', 'I do not love you!']
text_preprocessed = bert_preprocess_model(text_test)
text_preprocessed.keys()

dict_keys(['input_word_ids', 'input_mask', 'input_type_ids'])

In [5]:
text_preprocessed['input_mask']

<tf.Tensor: shape=(3, 128), dtype=int32, numpy=
array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0

In [6]:
text_preprocessed['input_type_ids']

<tf.Tensor: shape=(3, 128), dtype=int32, numpy=
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0

In [7]:

text_preprocessed['input_word_ids']

<tf.Tensor: shape=(3, 128), dtype=int32, numpy=
array([[  101,  2026,  2171,  2003,  2024,  2546, 14863,  2094,  5558,
         6977,   102,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0, 

**101 --> CLS token**

102 --> SEP token

BERT uses CLS as a special token at the begining of each setence whereas SEP as a special token to separate two sentences or end sinle sentece

In [8]:
bert_model = hub.KerasLayer(encoder_url)

In [9]:
bert_results = bert_model(text_preprocessed)

In [10]:
bert_results.keys()

dict_keys(['encoder_outputs', 'pooled_output', 'default', 'sequence_output'])

In [11]:
bert_results['sequence_output']

<tf.Tensor: shape=(3, 128, 768), dtype=float32, numpy=
array([[[ 0.03164041,  0.65168417, -0.29465505, ..., -0.07567722,
          0.10913479,  0.46711525],
        [ 0.12765312,  0.27567992, -0.43587252, ..., -0.04122534,
         -0.2781331 ,  0.21969342],
        [-0.34649527,  0.5120872 ,  0.26744062, ..., -0.2752505 ,
         -0.08428204,  0.05997363],
        ...,
        [-0.04670744,  0.56670123,  0.44889823, ...,  0.31258595,
          0.02284842, -0.03401123],
        [ 0.04515839,  0.5801153 ,  0.5187756 , ...,  0.2673342 ,
          0.01117608, -0.09117278],
        [-0.00777702,  0.56120473,  0.49309987, ...,  0.28632623,
          0.015908  , -0.11355118]],

       [[-0.11412742,  0.24770765,  0.18593174, ..., -0.3240056 ,
          0.15802622,  0.34969097],
        [ 0.43220595,  0.06715905,  0.14354464, ..., -0.64845854,
          0.4207252 , -0.15595299],
        [ 1.2384564 ,  0.58226615,  1.0079719 , ..., -0.31359917,
          0.46565297,  0.1319511 ],
        ...,

In [12]:
bert_results['pooled_output']

<tf.Tensor: shape=(3, 768), dtype=float32, numpy=
array([[-0.86282825, -0.4938773 , -0.6848537 , ..., -0.7438012 ,
        -0.7516517 ,  0.9084357 ],
       [-0.8977165 , -0.388722  , -0.6252476 , ..., -0.28664523,
        -0.6398827 ,  0.87938595],
       [-0.8176715 , -0.29529753, -0.498667  , ..., -0.05026734,
        -0.63220024,  0.8750386 ]], dtype=float32)>

In [13]:
len(bert_results['encoder_outputs'])

12

Since we are using BERT base model it has 12 encoder layers, that's why above the length of encoder_output is 12. 

Read more about purpose of each element here: https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4



You can see below that last element of encoder_outputs is basically a sequence_output

In [15]:
bert_results['encoder_outputs'][-1] == bert_results['sequence_output']

<tf.Tensor: shape=(3, 128, 768), dtype=bool, numpy=
array([[[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]],

       [[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]],

       [[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  

In [16]:
bert_results['encoder_outputs']

[<tf.Tensor: shape=(3, 128, 768), dtype=float32, numpy=
 array([[[ 1.18907966e-01,  1.17901891e-01, -9.85742658e-02, ...,
           3.67179513e-04,  9.23789069e-02,  8.09294879e-02],
         [ 8.54261875e-01,  3.74532640e-01, -7.11446822e-01, ...,
          -9.30342615e-01,  5.31657338e-01, -2.30339199e-01],
         [ 4.25793082e-01,  5.43034136e-01, -1.35024309e+00, ...,
          -8.37689698e-01,  1.97234362e-01, -1.16463292e+00],
         ...,
         [ 1.11245289e-02, -1.84480071e-01,  5.34858584e-01, ...,
           3.64159912e-01, -2.07879394e-01,  4.32990193e-02],
         [-8.47320482e-02, -1.92653939e-01,  4.33150500e-01, ...,
           3.66674364e-01, -1.37640148e-01, -5.14687225e-02],
         [ 5.78204021e-02, -1.14007145e-01,  4.55207199e-01, ...,
           6.84337556e-01, -4.48037267e-01, -1.21677361e-01]],
 
        [[ 1.54931009e-01,  1.07907966e-01, -7.27573484e-02, ...,
           2.73726285e-02,  1.04464926e-01,  3.94904539e-02],
         [ 4.80797380e-01,  7.0

**BERT tutorial: Classify spam vs no spam emails**


---



---





In [1]:
#pip install tensorflow-text

In [2]:
import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_text as text

Import the dataset (Dataset is taken from kaggle)


https://www.kaggle.com/datasets/venky73/spam-mails-dataset

In [3]:
import pandas as pd

df = pd.read_csv("/content/sample_data/spam_ham_dataset.csv")
df.head(5)

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [4]:
df.groupby('Category').describe()

Message                                                            \
           count unique                                                top   
Category                                                                     
ham         4825   4516                             Sorry, I'll call later   
spam         747    641  Please call our customer service representativ...   

               
         freq  
Category       
ham        30  
spam        4

In [5]:
df['spam']=df['Category'].apply(lambda x: 1 if x=='spam' else 0)
df.head()

Category                                            Message  spam
0      ham  Go until jurong point, crazy.. Available only ...     0
1      ham                      Ok lar... Joking wif u oni...     0
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...     1
3      ham  U dun say so early hor... U c already then say...     0
4      ham  Nah I don't think he goes to usf, he lives aro...     0

Split it into training and test data set

In [6]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df['Message'],df['spam'], stratify=df['spam'])

In [7]:
X_train.head(4)

3909    Sounds like a plan! Cardiff is still here and ...
565     Fuck babe ... I miss you already, you know ? C...
1515                         K:)all the best:)congrats...
3333    Ok... The theory test? when are ü going to boo...
Name: Message, dtype: object

Now lets import BERT model and get embeding vectors for few sample statements

In [8]:
bert_preprocess = hub.KerasLayer("https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3")
bert_encoder = hub.KerasLayer("https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4")

In [9]:
def get_sentence_embeding(sentences):
    preprocessed_text = bert_preprocess(sentences)
    return bert_encoder(preprocessed_text)['pooled_output']

get_sentence_embeding([
    "500$ discount. hurry up", 
    "Bhavin, are you up for a volleybal game tomorrow?"]
)

<tf.Tensor: shape=(2, 768), dtype=float32, numpy=
array([[-0.8435166 , -0.5132724 , -0.88845706, ..., -0.7474883 ,
        -0.7531471 ,  0.91964483],
       [-0.8720836 , -0.50544   , -0.9444667 , ..., -0.8584748 ,
        -0.71745366,  0.88082993]], dtype=float32)>

Get embeding vectors for few sample words. Compare them using cosine similarity

In [10]:
e = get_sentence_embeding([
    "banana", 
    "grapes",
    "mango",
    "jeff bezos",
    "elon musk",
    "bill gates"
]
)

In [11]:
from sklearn.metrics.pairwise import cosine_similarity
cosine_similarity([e[0]],[e[1]])

array([[0.99110895]], dtype=float32)

Values near to 1 means they are similar. 0 means they are very different. Above you can use comparing "banana" vs "grapes" you get 0.99 similarity as they both are fruits

In [12]:
cosine_similarity([e[0]],[e[3]])

array([[0.8470383]], dtype=float32)

Comparing banana with jeff bezos you still get 0.84 but it is not as close as 0.99 that we got with grapes

In [13]:
cosine_similarity([e[3]],[e[4]])

array([[0.9872036]], dtype=float32)

Jeff bezos and Elon musk are more similar then Jeff bezos and banana as indicated above

**Build Model**


There are two types of models you can build in tensorflow.

(1) Sequential (2) Functional

So far we have built sequential model. But below we will build functional model. 

More information on these two is here: https://becominghuman.ai/sequential-vs-functional-model-in-keras-20684f766057

In [14]:
# Bert layers
text_input = tf.keras.layers.Input(shape=(), dtype=tf.string, name='text')
preprocessed_text = bert_preprocess(text_input)
outputs = bert_encoder(preprocessed_text)

# Neural network layers
l = tf.keras.layers.Dropout(0.1, name="dropout")(outputs['pooled_output'])
l = tf.keras.layers.Dense(1, activation='sigmoid', name="output")(l)

# Use inputs and outputs to construct a final model
model = tf.keras.Model(inputs=[text_input], outputs = [l])

In [15]:
model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 text (InputLayer)              [(None,)]            0           []                               
                                                                                                  
 keras_layer (KerasLayer)       {'input_word_ids':   0           ['text[0][0]']                   
                                (None, 128),                                                      
                                 'input_type_ids':                                                
                                (None, 128),                                                      
                                 'input_mask': (Non                                               
                                e, 128)}                                                      

In [16]:
len(X_train)

4179

In [17]:
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

Train the model

In [18]:
model.fit(X_train, y_train, epochs=5)

Epoch 1/5
131/131 [==============================] - 90s 601ms/step - loss: 0.3457 - accuracy: 0.8571
Epoch 2/5
131/131 [==============================] - 80s 609ms/step - loss: 0.2520 - accuracy: 0.8837
Epoch 3/5
131/131 [==============================] - 80s 609ms/step - loss: 0.2096 - accuracy: 0.9124
Epoch 4/5
131/131 [==============================] - 80s 608ms/step - loss: 0.1845 - accuracy: 0.9282
Epoch 5/5
131/131 [==============================] - 80s 610ms/step - loss: 0.1669 - accuracy: 0.9409


In [19]:
model.evaluate(X_test, y_test)

44/44 [==============================] - 31s 677ms/step - loss: 0.1433 - accuracy: 0.9591


[0.14327920973300934, 0.9590811133384705]

Interface

In [20]:
reviews = [
    'Reply to win £100 weekly! Where will the 2006 FIFA World Cup be held? Send STOP to 87239 to end service',
    'You are awarded a SiPix Digital Camera! call 09061221061 from landline. Delivery within 28days. T Cs Box177. M221BP. 2yr warranty. 150ppm. 16 . p p£3.99',
    'it to 80488. Your 500 free text messages are valid until 31 December 2005.',
    'Hey Sam, Are you coming for a cricket game tomorrow',
    "Why don't you wait 'til at least wednesday to see if you get your ."
]
model.predict(reviews)

array([[0.67477167],
       [0.73162025],
       [0.629449  ],
       [0.076676  ],
       [0.02578088]], dtype=float32)

Near 1 means it is most likely to be spam.